synapseclient/client.py

```diff
@@ -2110,6 +2112,163 @@ def _uploadStringToFile(self, content, contentType="text/plain"):
+            'storageLocationId': storageLocationId
+        }
+
+        return DictObject(**self.restPOST(uri='/file/multipart',
+                                          body=json.dumps(upload_request),
+                                          endpoint=self.fileHandleEndpoint))
+
+
+    def _get_presigned_urls(self, uploadId, partNumbers):
+        """
+        :returns: A BatchPresignedUploadUrlResponse_.
+
+        .. BatchPresignedUploadUrlResponse: http://rest.synapse.org/POST/file/multipart/uploadId/presigned/url/batch.html
+        """
+        presigned_url_request = {
+            'uploadId':uploadId,
+            'partNumbers':partNumbers
+        }
+
+        uri = '/file/multipart/{uploadId}/presigned/url/batch'.format(uploadId=uploadId)
+        return DictObject(**self.restPOST(uri, json.dumps(presigned_url_request),
+                                          endpoint=self.fileHandleEndpoint))
+
+
+    def _add_part(self, uploadId, partNumber, partMD5Hex):
+        """
+        :returns: An AddPartResponse_ with fields for an errorMessage and addPartState containing
+                  either 'ADD_SUCCESS' or 'ADD_FAILED'.
+
+        .. AddPartResponse: http://rest.synapse.org/org/sagebionetworks/repo/model/file/AddPartResponse.html
+        """
+        uri = '/file/multipart/{uploadId}/add/{partNumber}?partMD5Hex={partMD5Hex}'.format(**locals())
+        return DictObject(**self.restPUT(uri, endpoint=self.fileHandleEndpoint))
+
+
+    def _complete_multipart_upload(self, uploadId):
+        """
+        :returns: A MultipartUploadStatus_.
+
+        .. MultipartUploadStatus: http://rest.synapse.org/org/sagebionetworks/repo/model/file/MultipartUploadStatus.html
+        """
+        uri = '/file/multipart/{uploadId}/complete'.format(uploadId=uploadId)
+        return DictObject(**self.restPUT(uri, endpoint=self.fileHandleEndpoint))
+
+
+    def multipartUpload(self, filepath, partSize=None, contentType=None):
+        PART_RETRIES = 5
+        URL_BATCH_SIZE = 12
+        MAX_PARTS = 10000
+        PART_RETRIES = 3
+
+        def find_parts_by_status(part_status, status='0'):
+            """
+            Given a string of the form "1001110", where 1 and 0 indicate a status of
+            completed or not, return the part numbers having the specified status.
+            """
+            return [i+1 for i,c in enumerate(part_status) if c==status]
+
+        def partition(n, list):
+            """
+            Split the input list into partitions of size n.
+            """
+            for i in range(0, len(list), n):
+                yield list[i:i+n]
+
+        if not os.path.exists(filepath):
+            raise IOError('File "%s" not found.' % filepath)
+        if os.path.isdir(filepath):
+            raise IOError('File "%s" is a directory.' % filepath)
+
+        fileSize = os.path.getsize(filepath)
+
+        if partSize is None:
+            partSize = max(5*MB, math.ceil(fileSize/float(MAX_PARTS)))
```

Comment:
Make this a separate function and test. Make sure we:
* can't get 10,001 parts
* can't get parts less than 5 MB